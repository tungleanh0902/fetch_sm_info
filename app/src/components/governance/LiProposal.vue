<template>
  <div
    class="li-proposal"
    @click="
      $router.push({
        name: 'proposal',
        params: { proposalId: String(proposal.id) },
      })
    "
  >
    <div class="proposal-content">
      <div>
        <div class="status-and-id">
          <Status :status="status" />
        </div>
        <h3 class="title">
          {{ proposal.title }}
        </h3>
      </div>
      <span v-if="proposal.creationTime" class="time">{{
        proposal.creationTime | moment
      }}</span>
      <span v-else class="time">n/a</span>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex"
import moment from "moment"
import { getProposalStatus } from "scripts/proposal-status"
import Status from "common/Status"

export default {
  name: `li-proposal`,
  components: {
    Status,
  },
  filters: {
    moment: function (date) {
      return moment(date).fromNow()
    },
  },
  props: {
    proposal: {
      type: Object,
      required: true,
    },
  },
  computed: {
    ...mapState([`proposals`]),
    status() {
      return getProposalStatus(this.proposal)
    },
  },
}
</script>

<style scoped>
.title {
  font-size: 22px;
  color: var(--bright);
  font-weight: 500;
  max-width: 600px;
}

.li-proposal {
  padding: 2rem 0;
  margin: 0 auto;
  display: block;
  cursor: pointer;
  max-width: 1024px;
  width: 100%;
  border-bottom: 2px solid var(--bc-dim);
}

.li-proposal:first-child {
  border-top: 2px solid var(--bc-dim);
}

.li-proposal:hover {
  cursor: pointer;
}

.li-proposal:hover .title {
  color: var(--link);
}

.proposal-content {
  display: flex;
  justify-content: space-between;
}

.status-and-id {
  display: flex;
  padding-bottom: 1rem;
}

.time {
  align-self: flex-end;
  color: var(--dim);
  font-size: 14px;
}

@media screen and (max-width: 1023px) {
  .proposal-content {
    flex-direction: column;
  }

  .time {
    align-self: auto;
    padding-top: 1rem;
  }
}
</style>
