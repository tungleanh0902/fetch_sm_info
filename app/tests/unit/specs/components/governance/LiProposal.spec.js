import { shallowMount, createLocalVue } from "@vue/test-utils"
import LiProposal from "src/components/governance/LiProposal"

import { proposals } from "../../store/json/proposals"

const proposal = proposals[`1`]

describe(`LiProposal`, () => {
  const localVue = createLocalVue()
  localVue.directive(`tooltip`, () => {})

  let wrapper

  beforeEach(() => {
    const $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
    }

    wrapper = shallowMount(LiProposal, {
      localVue,
      mocks: {
        $store,
      },
      propsData: { proposal },
      stubs: [`router-link`],
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should return status info for passed proposals`, () => {
    wrapper.setProps({
      proposal: {
        ...proposal,
        status: `Passed`,
      },
    })
    expect(wrapper.vm.status).toEqual({
      title: `Passed`,
      value: `PASSED`,
      color: `green`,
    })
  })

  it(`should return status info for rejected proposals`, () => {
    wrapper.setProps({
      proposal: {
        ...proposal,
        status: `Rejected`,
      },
    })
    expect(wrapper.vm.status).toEqual({
      title: `Rejected`,
      value: `REJECTED`,
      color: `red`,
    })
  })

  it(`should return status info for active proposals`, () => {
    wrapper.setProps({
      proposal: {
        ...proposal,
        status: `VotingPeriod`,
      },
    })
    expect(wrapper.vm.status).toEqual({
      title: `Voting Period`,
      value: `VOTING`,
      color: `highlight`,
    })
  })

  it(`should return status info for 'DepositPeriod' proposals`, () => {
    wrapper.setProps({
      proposal: {
        ...proposal,
        status: `DepositPeriod`,
      },
    })
    expect(wrapper.vm.status).toEqual({
      title: `Deposit Period`,
      value: `DEPOSITING`,
      color: `orange`,
    })
  })

  it(`should return status info for an unknown proposal type`, () => {
    wrapper.setProps({
      proposal: {
        ...proposal,
        status: `Unknown`,
      },
    })
    expect(wrapper.vm.status).toEqual({
      title: `Error`,
      value: `UNKNOWN`,
      color: `grey`,
    })
  })

  it(`should survive the tally result not being present yet`, () => {
    const $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      state: {
        proposals: {
          tallies: {},
        },
      },
    }

    wrapper = shallowMount(LiProposal, {
      localVue,
      mocks: {
        $store,
      },
      propsData: { proposal },
      stubs: [`router-link`],
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
