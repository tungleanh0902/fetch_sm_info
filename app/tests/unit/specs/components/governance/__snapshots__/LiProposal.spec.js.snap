// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiProposal has the expected html structure 1`] = `
<div
  class="li-proposal"
>
  <div
    class="proposal-content"
  >
    <div>
      <div
        class="status-and-id"
      >
        <status-stub
          status="[object Object]"
        />
      </div>
       
      <h3
        class="title"
      >
        
        A new test proposal
      
      </h3>
    </div>
     
    <span
      class="time"
    >
      in 50 years
    </span>
  </div>
</div>
`;

exports[`LiProposal should survive the tally result not being present yet 1`] = `
<div
  class="li-proposal"
>
  <div
    class="proposal-content"
  >
    <div>
      <div
        class="status-and-id"
      >
        <status-stub
          status="[object Object]"
        />
      </div>
       
      <h3
        class="title"
      >
        
        A new test proposal
      
      </h3>
    </div>
     
    <span
      class="time"
    >
      in 50 years
    </span>
  </div>
</div>
`;
